extensions/parquet/table/src/main/java/io/deephaven/parquet/table/metadata/RowGroupInfo.java

```diff
@@ -0,0 +1,324 @@
+                }
+            }
+
+            if (it.hasMore()) {
+                try (final RowSet leftoverRows = it.getNextRowSequenceWithLength(Long.MAX_VALUE).asRowSet()) {
+                    throw new IllegalStateException(String.format("Subtable dropped rows; Iterator has more: %s",
+                            leftoverRows));
+                }
+            }
+        }
+
+        return ret;
+    }
+     */
+    // @formatter:on
+
+    /**
+     * Keeps all rows within a single RowGroup
+     */
+    private static class SingleRowGroup extends RowGroupInfo {
+        @Override
+        public Table[] splitForRowGroups(@NotNull final Table input) {
+            return new Table[] {input};
+        }
+    }
+
+    /**
+     * Splits evenly across a number of RowGroups, ensuring that no group is larger than {@code maxRows}
+     */
+    private static class SplitByMaxRows extends RowGroupInfo {
+        private final long maxRows;
+
+        private SplitByMaxRows(long maxRows) {
+            if (maxRows <= 0) {
+                throw new IllegalArgumentException("MaxRows must be positive");
+            }
+            this.maxRows = maxRows;
+        }
+
+        @Override
+        public Table[] splitForRowGroups(@NotNull final Table input) {
+            final long numRowGroups = (input.size() / (maxRows - 1)) + 1;
+            if (numRowGroups > Integer.MAX_VALUE) {
+                throw new IllegalStateException("Number of RowGroups is greater than Integer.MAX_VALUE");
+            }
+            return splitEvenly(input, (int) numRowGroups);
+        }
+    }
+
+    /**
+     * Splits evenly across {@code numRowGroups} RowGroups
+     */
+    private static class SplitEvenly extends RowGroupInfo {
+        private final int numRowGroups;
+
+        private SplitEvenly(int numRowGroups) {
+            this.numRowGroups = numRowGroups;
+        }
+
+        @Override
+        public Table[] splitForRowGroups(@NotNull final Table input) {
+            return splitEvenly(input, numRowGroups);
+        }
+    }
+
+    /**
+     * Splits into RowGroups based on unique values within {@code groups}, and optionally further splits (per unique
+     * {@code groups} ensuring that no group is larger than {@code maxRows}. If {@code maxRows} is not desired, this
+     * parameter may be set to {@code Long.MAX_VALUE}
+     */
+    private static class SplitByGroups extends CustomSplit {
+        private SplitByGroups(final long maxRows, final String[] groups) {
+            super(table -> splitByGroups(table, maxRows, groups));
+            if (groups == null || groups.length == 0) {
+                throw new IllegalArgumentException("Requires at least one group");
+            }
+        }
+
+        private static Table[] splitByGroups(@NotNull final Table input, final long maxRows, final String[] groups) {
+            final Table[] partitionedTables = input.partitionBy(groups).constituents();
```

Comment:
Even though a properly ordered input Table might be passed in here, I don't _think_ we are guaranteed that the partitionBy outputs will share the same rowset keyspace.